### PR TITLE
Add Vector2.from_angle() method

### PR DESCRIPTION
**Pull request by KoBeWi**
Resolves 2D part of https://github.com/godotengine/godot-proposals/issues/1719 (someone wanted Vector3 equivalent, but it probably requires some more math so meh)

With this you can do
```
Vector2.from_angle(PI/2)
```
and it will return `Vector2(0, 1)`

(or at least it would if static functions weren't bugged, see #48238)

### CODE REVIEW

File: doc/classes/Vector2.xml
Comment (kleonc) [inline doc/classes/Vector2.xml:199]: `Vector2.from_angle(angle).angle()` will return the `angle` wrapped into the `[-PI, PI]`  range so it might not be equal to the `angle`. Probably docs of `@GDscript.atan2`/`Vector2.angle` should mention their results are in `[-PI, PI]` range too.
It's also worth mentioning the `angle` is in radians.

Comment (KoBeWi) [reply doc/classes/Vector2.xml:199]: Well, they are equal assuming that PI equals to -PI (as this is equivalent vector). I should probably reword it 🙃

Comment (KoBeWi) [reply doc/classes/Vector2.xml]: Ok changed. I removed the `angle()` mention and just added a code sample.

File: core/math/vector2.cpp
Comment (Xrayez) [inline core/math/vector2.cpp:37]: Maybe:

```suggestion
Vector2 Vector2::from_angle(real_t p_radians) {
```

since `from_angle` already implies it,

File: doc/classes/Vector2.xml
Comment (Xrayez) [inline doc/classes/Vector2.xml:203]: Missing parenthesis:

```suggestion
				print(Vector2.from_angle(0)) # Prints Vector2(1, 0).
				print(Vector2(1, 0).angle()) # Prints 0, which is the angle used above.
				print(Vector2.from_angle(PI/2)) # Prints Vector2(0, 1).
```

**Review by Xrayez — CHANGES_REQUESTED** [review]

File: doc/classes/Vector2.xml
Comment (aaronfranke) [inline doc/classes/Vector2.xml:203]:
```suggestion
				print(Vector2.from_angle(TAU / 4)) # Prints Vector2(0, 1).
```

File: core/math/vector2.h
Comment (aaronfranke) [inline core/math/vector2.h:131]: I think the argument can be made const.

File: doc/classes/Vector2.xml
Comment (LightningAA) [inline doc/classes/Vector2.xml:199]: Nitpicking, but should be "equivalent **to**"

File: doc/classes/Vector2.xml
Comment (Xrayez) [reply doc/classes/Vector2.xml:203]: Why `TAU`? In both cases this produces `Vector2(-0, 1)` from my testing. Also `PI / 2` and similar is more common to see in math in comparison.

Comment (LightningAA) [reply doc/classes/Vector2.xml:203]: I like TAU more for angles because TAU / 4.0 is one fourth of a circle, which makes more sense to me than PI / 2.0 which is half of half a circle, although you're right it's just a preference thing. Also according to the [GDScript style guide](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_styleguide.html#numbers) these numbers should have a `.0` at the end.

Comment (aaronfranke) [reply doc/classes/Vector2.xml:203]: That's not what the GDScript style guide says (although I'm not opposed to this change). It says that floats should have `x.0` or `0.x` instead of just `x.` or `.x`. In this case, `TAU` is already a float, so in the expression `TAU / 4`, `4` is not technically a float, it's an integer which is implicitly converted to a float. According to the style guide, either `TAU / 4` or `TAU / 4.0` is appropriate. That said, `TAU / 4.0` is more explicit and likely better.

Comment (KoBeWi) [reply doc/classes/Vector2.xml]: PI is more recognizable, so I'm leaving it.

**Review by aaronfranke — CHANGES_REQUESTED** [review]
Doesn't work in my testing:

Screenshot from 2021-04-27 17-26-57

Missing test cases.

Missing C# version.

**Comment by Xrayez** [comment]
> Missing test cases.

Writing tests is not currently required (unlike documentation), but highly recommended. See also godotengine/godot-proposals#1586.

> Missing C# version.

Are there any contribution guides in the documentation to make this happen? I understand that C# is officially supported language, but I'm not sure if it's required for contributors to implement those methods currently. This question also applies to VisualScript, GDNative etc...

As for myself, I'm not using C# at all. Does that mean my own contribution to core would be blocked because of lack of C# knowledge?

**Comment by aaronfranke** [comment]
@Xrayez It's not necessarily a blocker, but if it's not done here, someone will have to do it later for consistency. It's easier to just do it everywhere at once.

As for VisualScript and GDNative, they get access to methods via `variant_call.cpp` (C# doesn't do this). I don't know if this includes static methods, but looking at the code, I don't see any special logic for from_hsv...

**Comment by KoBeWi** [comment]
> Doesn't work in my testing:

This is due to #48238.
You need to call it on a vector for now.

> Missing test cases.

This is task for #47202

> Missing C# version.

Added, but I never compiled mono version, so I can't test it.

**Comment by Xrayez** [comment]
> It's not necessarily a blocker, but if it's not done here, someone will have to do it later for consistency. It's easier to just do it everywhere at once.

Most of the time (if not always), I'd like to test a feature myself before doing changes. If I don't know how to test something, I'd rather avoid doing those changes, unless someone else is willing to test those changes before merging, otherwise this could lead to broken code.

I'd just make it clear that you're willing to test the implementation if you do want to see them in any PR, not just "Missing C# version". I'm just providing my feedback on the review process, sorry. 🙂

**Review by aaronfranke — APPROVED** [review]
The C# code looks good, and the tests can wait. This PR isn't fully functional until #48238 is fixed, so we can either fix that first, or merge this first and use this method as a test for #48238.

**Review by Xrayez — APPROVED** [review]

**Comment by groud** [comment]
The idea makes sense, but what you are providing here is simply half of of what would be needed to allow building any vector from [polar coordinates](https://en.wikipedia.org/wiki/Polar_coordinate_system).

Consequently, I think it would be more feature-complete to rename the function to `from_polar(...)` and to allow a second argument (that could be 1.0 by default), that would allowing providing a length too.

**Comment by KoBeWi** [comment]
The length argument wouldn't be very useful. It's a unit vector, so you can just multiply it.

Also the function was never intended to have anything to do with polar coordinates (which very few people use). `from_angle()` has similarity to `angle()` method, which makes it more obvious that the two methods are related, so it's easier to find.

**Comment by groud** [comment]
> The length argument wouldn't be very useful. It's a unit vector, so you can just multiply it. 
Also the function was never intended to have anything to do with polar coordinates (which very few people use). from_angle() has similarity to angle() method, which makes it more obvious that the two methods are related, so it's easier to find.

I mean, at least it looks like something that has a mathematical sense. The `angle()` function itself is basically returning a float out of a vector, without the 3 points usually needed to define an angle. This is simply because it returns the angle part of it's polar coordinates.

I think that, if we implement the possibility to create a vector from an angle, then that's basically equivalent to building a vector from it's polar coordinates (returned by the angle() and lenght() function the other way around). It makes the API more consistent and easier to search.

**Comment by aaronfranke** [comment]
> without the 3 points usually needed to define an angle.

We don't need 3 points. The origin and X axis are implied.

When it comes to polar coordinates, this is just something that is super easy to do with existing methods (angle & length, and from_angle & multiply), while simultaneously being a very niche use case for games, so it doesn't make sense to have APIs for it.

**Comment by groud** [comment]
> The origin and X axis are implied.

Yep, that's basically the issue here. As explicit naming is better than implicit naming, I think `from_polar` is better than `from_angle()` as it is more explicit.

**Comment by Xrayez** [comment]
Compare:
```gdscript
Vector2.from_angle(radians) * length
Vector2.from_polar(radians, length)
```
I really do think it's not worth being explicit here.

@groud your opinion on this is basically the complete opposite of what you said in https://github.com/godotengine/godot/pull/43103#issuecomment-717276143:

> Simply allow generating a random unit vector then let the user multiply it. That's a simple and efficient API.

The `from_angle` is "simple and efficient" API as well according to your comment. I understand that I'm taking this from other context, but it's very similar to what we have here.

**Comment by Xrayez** [comment]
Alternative solution could be:
1. Merge this PR as-is;
2. Implement new methods such as:
   - `Geometry2D.polar_to_vector(theta, length)`
   - `Geometry3D.polar_to_vector(theta, phi, length)`

if really needed. Due to this, I've even considered to do the same thing in goostengine/goost#65, where using `polar` terminology would probably make more sense, since `Geometry` already implies that we deal with something related to *space*, contrary to `Vector2/3` classes which are not always used in a spatial context.

> Consequently, I think it would be more feature-complete to rename the function to `from_polar(...)` and to allow a second argument (that could be 1.0 by default), that would allowing providing a length too.

Being feature-complete can mean anything to a user. For instance, are we going to implement other conversion methods such [cylindrical coordinate system](https://en.wikipedia.org/wiki/Cylindrical_coordinate_system) as well? That would certainly make those features complete, but I'm afraid those methods could bloat `Vector2/3` classes with features that nobody is going to use in their *game* projects.

That's why I'm suggesting that we implement coordinate system conversion methods in classes such as `Geometry2D/3D` (again, if *needed*), and merge this PR for the reasons already outlined above by @KoBeWi and @aaronfranke.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
ready to merge when passes CI. Do not forget to remove polar/cartesian functions from variant_utility in a subsequent PR, as they are redundant and a confusing API for Godot users compared to this one.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs:614]:
```suggestion
	          return new Vector2(Mathf.Cos(angle), Mathf.Sin(angle));
```

You only have this line intended by 2 spaces instead of 4.

**Review by aaronfranke — COMMENTED** [review]
Works in my testing and looks good except for one style issue.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs
Comment (KoBeWi) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs:614]: Shouldn't it fail static checks? clang-format auto-fixes this 🤔

**Review by aaronfranke — APPROVED** [review]
Looks good and works in my testing. I haven't tested the C# version but it looks good.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs
Comment (aaronfranke) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Vector2.cs]: I don't think we have any checks that apply to C#, except for the trivial `file_format.sh`. We should add some checks for the formatting of these files eventually.

EDIT: The repo now has `dotnet-format`.

**Comment by vnen** [comment]
Thanks!

**Pull request merged**
